inc/t_cose/t_cose_mac_verify.h

```diff
@@ -0,0 +1,123 @@
+ * t_cose_sign1_get_nth_tag(). The CWT
+ * tag is an example of the tags that might returned. The COSE tags
+ * that are processed, don't count here.
+ */
+#define T_COSE_MAX_TAGS_TO_RETURN 4
+
+/**
+ * Context for tag verification.  It is about 24 bytes on a
+ * 64-bit machine and 12 bytes on a 32-bit machine.
+ */
+struct t_cose_mac_verify_ctx {
+    /* Private data structure */
+    struct t_cose_key     verification_key;
+    int32_t               option_flags;
+    uint64_t              auTags[T_COSE_MAX_TAGS_TO_RETURN];
+};
+
+
+/**
+ * \brief Initialize for \c COSE_Mac0 message verification.
+ *
+ * \param[in,out]  context       The context to initialize.
+ * \param[in]      option_flags  Options controlling the verification.
+ *
+ * This must be called before using the verification context.
+ */
+static void
+t_cose_mac_verify_init(struct t_cose_mac_verify_ctx *context,
+                        int32_t                      option_flags);
+
+
+/**
+ * \brief Set key for \c COSE_Mac0 message verification.
+ *
+ * \param[in,out] context      The context of COSE_Mac0 verification
+ * \param[in] verify_key       The verification key to use.
+ *
+ * Look up by kid parameter and fetch the key for MAC verification.
+ * Setup the \ref verify_key structure and fill it in \ref context.
+ */
+static void
+t_cose_mac_set_verify_key(struct t_cose_mac_verify_ctx *context,
+                           struct t_cose_key            verify_key);
+
+/**
+ * \brief Verify a COSE_Mac0
+ *
+ * \param[in] context      The context of COSE_Mac0 verification
+ * \param[in] cose_mac    Pointer and length of CBOR encoded \c COSE_Mac0
+ *                         that is to be verified.
+ * \param[out] payload     Pointer and length of the still CBOR encoded
+ *                         payload
+ *
+ * \return This returns one of the error codes defined by \ref t_cose_err_t.
+ *
+ * Verification involves the following steps.
+ *
+ * The CBOR structure is parsed and verified. It makes sure \c COSE_Mac0
+ * is valid CBOR and that it is tagged as a \c COSE_Mac0.
+ *
+ * The signing algorithm is pulled out of the protected headers.
+ *
+ * The kid (key ID) is parsed out of the unprotected headers if it exists.
+ *
+ * The payload is identified. It doesn't have to be parsed in detail
+ * because it is wrapped in a bstr.
+ *
+ * Finally, the MAC verification is performed if \ref T_COSE_OPT_DECODE_ONLY
+ * is not set in option flag. Otherwise, the verification will be skipped.
+ * The MAC algorithm to use comes from the signing algorithm in the
+ * protected headers.
+ * If the algorithm is not known or not supported this will error out.
+ *
+ * If it is successful, the pointer of the CBOR-encoded payload is
+ * returned.
+ */
+enum t_cose_err_t t_cose_mac_verify(struct t_cose_mac_verify_ctx *context,
+                                     struct q_useful_buf_c        cose_mac,
+                                     struct q_useful_buf_c       *payload);
```

Comment:
I had way more debug related issues than I expected which is the reason that resolving this stuff took so long I managed to resolve everything but this since this part will require changing the mac ctx struct which I exepect will cascade into lots of more changes. For now feel free to review what I did for now.